Feat: BitTorrent v2 Support [BEP 52]

leoherzog opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix
[ X ] New feature
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

Added parsing of `infoHash` _and/or_ `infoHashv2`, as well as generation of v1, v2, or Hybrid torrents.

**Which issue (if any) does this pull request address?**

Issue #88

sneakers-the-rat commented on index.js:
```diff
@@ -11,31 +11,44 @@ import queueMicrotask from 'queue-microtask'
  * Parse a torrent identifier (magnet uri, .torrent file, info hash)
  * @param  {string|ArrayBufferView|Object} torrentId
+ * @param  {Object} options
+ * @param  {string} options.hashMode - 'v1', 'v2', or 'both' (default: 'v1')
```
Yes I think we would need to - wouldn't know in advance what version a torrent is at the time options are passed.

It's a three way switch, most reliably detected ime by checking for (length || files) for v1, file tree for v2, and both for hybrid. (Checking the meta version can be iffy and doesnt disambiguate hybrid from v2-only) :)

leoherzog replied:
@ThaUnknown I considered this. I suppose if the user only supplies an `infoHash` without `infoHashV2`, a `btih` without a `btmh`, or a torrent file with length or files without a file tree, then assume v1. Same for vice-versa.

Applied this change in https://github.com/webtorrent/parse-torrent/pull/193/commits/f34b051cbf824a3e9b068d204cdfe75450d0e13b

sneakers-the-rat commented on index.js:
```diff
@@ -11,31 +11,44 @@ import queueMicrotask from 'queue-microtask'
     return magnet(`magnet:?xt=urn:btih:${arr2hex(torrentId)}`)
+  } else if (ArrayBuffer.isView(torrentId) && torrentId.length === 32) {
+    // if info hash v2 (buffer)
+    return magnet(`magnet:?xt=urn:btmh:1220${arr2hex(torrentId)}`)
```
I'm not exactly sure how the torrentId field is used downstream, but I see ability to specify a torrent by SHA1 hash, SHA256 hash, but I'm not sure how we would pass both in the case of a hybrid torrent here. Would we just treat both swarms as separate? It is legal to have both `btih` and `btmh` in a magnet (and how hybrid magnets usually work)

sneakers-the-rat commented on index.js:
```diff
@@ -123,15 +138,24 @@ async function decodeTorrentFile (torrent) {
+
+  if (isV2) {
+    // BitTorrent v2 specific validation
+    ensure(torrent.info['file tree'], 'info[\'file tree\']')
```
Probably also check for presence of piece layers here too :)

leoherzog replied:
Great call. https://github.com/webtorrent/parse-torrent/pull/193/commits/5577ce60c3682741d39bd64b297a6e402aee91d2

sneakers-the-rat commented on index.js:
```diff
@@ -123,15 +138,24 @@ async function decodeTorrentFile (torrent) {
+    ensure(torrent.info['file tree'], 'info[\'file tree\']')
   } else {
-    ensure(typeof torrent.info.length === 'number', 'info.length')
+    // BitTorrent v1 validation
```
I am reading on phone so not sure if I'm getting the control flow right, but is this v1 validation leg exclusive with v2? We would want to do both in the case of hybrid torrents.

sneakers-the-rat commented on index.js:
```diff
@@ -175,6 +209,28 @@ async function decodeTorrentFile (torrent) {
+      }
+    }
+    processFileTree(torrent.info['file tree'])
+    torrent.info.files = files
```
Nice. Flattening the tree for easier processing is the right call, but it seems like we may be implicitly making a hybrid torrent here? Adding `files` without `pieces` would look like an (invalid) torrent file to some consumers. Thinking about what we might need to make piece validation easier downstream, and some method for flattening the file tree to get the path, length, and piece root would be good, but I think we may not want to assign it to the v1 files key

sneakers-the-rat commented:
Thanks for getting to this so quick, my week filled up faster than in anticipated. We will probably need to follow this up with some convenience methods for getting piece roots and piece layers at file offsets when working on the downstream webtorrent impl, but it looks like most of the parts we need are in place here, modulo open comments

leoherzog commented on index.js:
```diff
@@ -11,31 +11,44 @@ import queueMicrotask from 'queue-microtask'
-    // if info hash (buffer)
+    // if info hash v1 (buffer)
     return magnet(`magnet:?xt=urn:btih:${arr2hex(torrentId)}`)
+  } else if (ArrayBuffer.isView(torrentId) && torrentId.length === 32) {
```
https://github.com/webtorrent/parse-torrent/pull/193/commits/20374a9b16b6bcb978f19b97f3b3ee423d3c1673

Good call. Redundant. This better?

leoherzog commented:
@ThaUnknown @sneakers-the-rat Incorporated your feedback and made a bunch of tweaks. Can rebase/squash before merging, if you'd like.

leoherzog commented:
@ThaUnknown @sneakers-the-rat Awaiting feedback

ThaUnknown commented:
I'm aware I'm holding this back, the PR itself for the most part looks fine, but I don't know about the functionality, since I'd need to go into the specifics of btv2, read the spec, and test it myself, for which I don't have time as of now

leoherzog commented:
@ThaUnknown Cool. Let me know if I can be helpful in that somehow.

https://blog.libtorrent.org/2020/09/bittorrent-v2/

sneakers-the-rat commented:
Sorry dropped off my radar. I know the spec and have written a parser and generator so can evaluate. Marking unread to check out tmrw

sneakers-the-rat commented on index.js:
```diff
@@ -123,15 +138,29 @@ async function decodeTorrentFile (torrent) {
+  const hasV2Structure = !!torrent.info['file tree']
+
+  // BitTorrent v2 validation (when v2 structures present)
+  if (isV2 || hasV2Structure) {
```
i am not sure what degree of validation we want here, it seems like the previous implementation was just checking for presence of required keys rather than total validity so this is probably fine, but ya there is a lot more validation that could be done to e.g. ensure that the piece layers and file tree are constructed correctly. i am going to assume that this package leaves validation as a downstream concern

sneakers-the-rat commented on index.js:
```diff
@@ -123,15 +138,29 @@ async function decodeTorrentFile (torrent) {
-    ensure(typeof torrent.info.length === 'number', 'info.length')
+  const isV2 = torrent.info['meta version'] === 2
+  const hasV1Structure = !!(torrent.info.pieces || torrent.info.files || typeof torrent.info.length === 'number')
+  const hasV2Structure = !!torrent.info['file tree']
```
i'm assuming that the `ensure` calls below throw if something isn't present? if so, this could be simplified to just `hasV1Structure` and `hasV2Structure` and ensuring that `meta version` is present and equals 2 in the validation block below. having multiple vars that indicate whether something is v2 seems to complicate things below.

leoherzog replied:
Totally. Simplified this in 13c70e7.

sneakers-the-rat commented on index.js:
```diff
@@ -141,8 +170,21 @@ async function decodeTorrentFile (torrent) {
+  if (shouldGenerateV2) {
+    result.infoHashV2Buffer = await hash(result.infoBuffer, undefined, 'sha-256')
+    result.infoHashV2 = arr2hex(result.infoHashV2Buffer)
+  }
```
along with https://github.com/webtorrent/parse-torrent/pull/193/files#r2579332259 , this could be simplified 
```suggestion
  if (hasV1Structure) {
    result.infoHashBuffer = await hash(result.infoBuffer)
    result.infoHash = arr2hex(result.infoHashBuffer)
  }

  if (hasV2Structure) {
    result.infoHashV2Buffer = await hash(result.infoBuffer, undefined, 'sha-256')
    result.infoHashV2 = arr2hex(result.infoHashV2Buffer)
  }
```

leoherzog replied:
> Totally. Simplified this in 13c70e7.

Just like with Lines 142 to 144 comment above.

sneakers-the-rat commented on index.js:
```diff
@@ -176,7 +218,29 @@ async function decodeTorrentFile (torrent) {
+  if (hasV2Structure && !hasV1Structure) {
+    // Pure v2: flatten file tree for result.files (don't modify torrent.info.files)
+    files = []
+    function processFileTree (tree, currentPath = []) {
```
i would pull this out to a top-level function, this turns out to be necessary in a bunch of places. 

Personally, I would call this `flattenFileTree` because the inverse operation is also frequently needed (`unflattenFileTree`)

leoherzog replied:
Pulled it out to a top-level function :+1: 

a7771c3

sneakers-the-rat commented on index.js:
```diff
@@ -176,7 +218,29 @@ async function decodeTorrentFile (torrent) {
+    function processFileTree (tree, currentPath = []) {
+      for (const [name, entry] of Object.entries(tree)) {
+        const fullPath = [...currentPath, name]
+        if (entry.length !== undefined) {
```
i sort of hate that this is the technically correct way to identify if something is a file, but it is indeed what is spec'd in the v2 spec.

leoherzog replied:
Agreed. Messy side-effect of the spec. At least it was pulled out into the `flattenFileTree` function in a7771c3

sneakers-the-rat commented on index.js:
```diff
@@ -176,7 +218,29 @@ async function decodeTorrentFile (torrent) {
+            'path.utf-8': fullPath,
+            length: entry.length
+          })
+        } else {
```
i am not sure if this package handles [BEP-47](https://www.bittorrent.org/beps/bep_0047.html) file attributes, but they are pretty important for hybrid torrents, specifically for padfiles. maybe we could just spread the entry object here?

sneakers-the-rat replied:
this would be my main requested change^

leoherzog replied:
Something like this? 2ad638f

sneakers-the-rat replied:
yep, exactly

sneakers-the-rat commented on index.js:
```diff
@@ -194,7 +258,14 @@ async function decodeTorrentFile (torrent) {
+  } else {
+    // For v2 torrents without pieces, create empty array
+    result.pieces = []
+  }
```
I would not do this: this makes it ambiguous whether the torrent is improperly created (it actually has an empty pieces string) or just a v2-only torrent. imo `pieces` should remain unset for v2 torrents, and downstream code that depends on it being present should be forced to upgrade if they are handling v2-only torrents: previously this function would just throw for v2-only torrents, but now it doesn't, and so downstream code should be forced to throw if it doesn't support v2-only torrents.

leoherzog replied:
Good call. d9204a7

sneakers-the-rat commented on test/bittorrent-v2.js:
```diff
@@ -0,0 +1,93 @@
+  // magnet uri with v2 hash (btmh)
+  const magnetV2 = `magnet:?xt=urn:btmh:1220${v2Hash}`
+  parsed = await parseTorrent(magnetV2)
+  t.ok(parsed.infoHashV2)
```
why not the same assert equal in all of the cases?

leoherzog replied:
Cleaned up testing in 38f7f43

sneakers-the-rat commented on test/bittorrent-v2.js:
```diff
@@ -0,0 +1,93 @@
+  // v2 info hash (as a hex string - 64 characters)
+  const v2Hash = 'a'.repeat(64)
+  parsed = await parseTorrent(v2Hash)
+  t.equal(parsed.infoHashV2, v2Hash.toLowerCase())
```
probably also asserting that `infoHash` is undefined right?

leoherzog replied:
Yeah. Cleaned up testing in 38f7f43.

sneakers-the-rat commented on test/bittorrent-v2.js:
```diff
@@ -0,0 +1,93 @@
+  t.equal(parsed.infoHashV2, v2Hash.toLowerCase())
+
+  t.end()
+})
```
should also have a hybrid magnet link (`magnet:?xt=urn:btmh:...&urn:btih:...`) and ensure that both infohashes are parsed from it

leoherzog replied:
Makes sense. Added in 7903819.

sneakers-the-rat commented on test/bittorrent-v2.js:
```diff
@@ -0,0 +1,93 @@
+  // Test v2 torrent (should auto-detect and generate v2 hash)
+  const v2Parsed = await parseTorrent(v2Buf)
+  t.ok(v2Parsed.infoHashV2, 'v2 torrent should have v2 hash')
+  t.equal(v2Parsed.infoHashV2.length, 64, 'v2 hash should be 64 chars')
```
we should also assert known infohashes for these - correctness is important too!

leoherzog replied:
Cleaned up testing in 38f7f43 and dfb54c8

sneakers-the-rat commented on test/bittorrent-v2.js:
```diff
@@ -0,0 +1,93 @@
+
+  // Test that v2 torrent auto-detects and generates appropriate hashes
+  const parsed = await parseTorrent(torrentBuf)
+  t.ok(parsed.infoHashV2, 'v2 torrent should auto-generate v2 hash')
```
assert the infohash is correct

leoherzog replied:
Done. 38f7f43

sneakers-the-rat commented on test/bittorrent-v2.js:
```diff
@@ -0,0 +1,93 @@
+  // Test that v2 torrent auto-detects and generates appropriate hashes
+  const parsed = await parseTorrent(torrentBuf)
+  t.ok(parsed.infoHashV2, 'v2 torrent should auto-generate v2 hash')
+
```
also probably assert that v1-only things are not present (v1 infohash, pieces, etc.)

leoherzog replied:
Oh, right. dfb54c8

sneakers-the-rat reviewed (commented):
ok cool, this is much smaller than i thought it would be - this pretty much covers it if the goal is just to load it into a js object rather than validating.

can't really comment on compatibility with downstream use, but flatting the file tree is sort of the main thing, and downstream consumers will have to figure out how to deal with the absence of v1 `pieces` and validate with the file trees instead. 

one thing that might be useful is to have an additional `version` field that is `"v1" | "v2" | "hybrid"` so that downstream consumers can check that without needing to repeat the sort of awkward checks for presence of different fields

leoherzog committed:
Simplify v1/v2/Hybrid Detection

leoherzog committed:
Extract flattenFileTree to Top-Level Function

leoherzog committed:
Support BEP-47 File Attributes in flattenFileTree

leoherzog committed:
Don't Set pieces to Empty Array for v2 Torrents

leoherzog committed:
Add version Field

leoherzog committed:
Assert infoHash and pieces Undefined for v2-only Torrents

leoherzog committed:
Use Consistent Assertions and Verify Known Hashes

leoherzog committed:
Add Hybrid Magnet Link Test

leoherzog commented:
Thanks for looking at this, @sneakers-the-rat. I really appreciate your time and thoughts.

I _think_ I've addressed all of the raised points. It's a lot cleaner now. Again, I'll leave the commits all separate for clarity in this PR for now, but I can merge/squash into a couple of clearer commits when we think everything's resolved.

Also, I'm not sure if it's you maintainers' job to "resolve conversation" on the comment threads, or mine :)

sneakers-the-rat reviewed (approved):
(fwiw for maintainers, all my comments addressed :)

parthibx24 commented:
why is this stale 😭

leoherzog commented:
@ThaUnknown I have "resolved" all of the open comments from @sneakers-the-rat

ThaUnknown commented:
yes but webtorrent is dead until @feross fixes the NPM keys that have expired, so this won't be published until that's done anyways

my general lack of interest, or deep understanding of the protocol doesnt help, neither does the utter hatered for how horrificly this package is structured and written

sneakers-the-rat commented:
so where does that leave us? surely it can't take too much time for @feross to fix the keys right? this little bitty thing is just the first step in v2 support, so should we take this as a sign not to bother and treat webtorrent as abandoned? or do we merge this and continue even if we have feeling about the code structure?

leoherzog committed:
Add BitTorrent v2 Support with Dual Hash Validation

Squash of initial v2 implementation history (3289039..559f9f9), rebased
onto upstream master v11.0.21.

leoherzog committed:
Fix File Names and Single-File Paths When Flattening v2 File Trees

BEP 52 stores file properties under a zero-length key, which was being
appended as a path component, leaving every v2 file with an empty name.
Also detect single-file mode (the tree's only key repeats the torrent
name) so paths don't come out as name/name.

leoherzog committed:
Validate Torrent Structure and Round-Trip piece layers

- Restore the upstream error for strings that aren't a magnet, info
  hash, or file path (the refactored if/else chain silently returned
  undefined for unmatched strings)
- Reject torrents with neither v1 nor v2 structure instead of producing
  an object with no hashes and NaN length
- Check meta version per BEP 52 before trusting a file tree
- Carry piece layers through parsing so toTorrentFile produces a valid
  v2/hybrid torrent file instead of one that fails re-parsing

ThaUnknown commented:
diego fixed it, but I simply dont have the time, knowledge, experience or interest to work on this right now, sorry

leoherzog commented:
Are there other maintainers of this project?

sneakers-the-rat commented:
i've reviewed this, it is backwards compatible, just adding support for v2. @ThaUnknown totally understood if you don't have the time, but i do know the v2 spec and can confirm this adds support at least for parsing v2 torrents, is there any way we can help you out so we can get support for v2 in webtorrent rolling without needing you to spend a bunch of time on it? 

this part was just supposed to be a super low-hanging fruit stepping stone towards support & is very low impact, v2 would be a big asset for webtorrent in particular to be able to serve web content, since per-file hashes make it possible for clients to be smart about deduplicating identical files like fonts, minified js, and etc. (our goal is to use webtorrent for a bt-backed wayback machine), but if we can't get this off the ground then it would be good to know that webtorrent is the wrong tech for that project so we can redirect that work elsewhere.
